### PR TITLE
Übersicht Feedbacks / Stundenerfassungen

### PR DESCRIPTION
**Pull request by casaper**
# [Story in Trello](https://trello.com/c/YI4RW6mk/11-8-%C3%BCbersicht-feedbacks-stundenerfassungen)

### Done for approval?

* [x] Acceptance criteria are met
* [x] Code quality checks are green
* [x] Readme updated if needed
* [x] Story under test
* [x] Mobile works
* [x] Seeds created
* [ ] Code is reviewed

### Done after the merge?

* [ ] Deployed to staging after the merge
* [ ] Tested manually on staging

### CODE REVIEW

File: app/controllers/list_responses_controller.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
+class ListResponsesController < ApplicationController
+  def feedbacks
+    authorize :list_response, :feedbacks?
+    @feedbacks = Feedback.order(created_at: :asc).where(marked_done_by_id: nil).paginate(page: params[:page])
```
Comment (toupeira) [inline app/controllers/list_responses_controller.rb]: `Feedback.created_asc.where(...` (was recently added to `ApplicationRecord`)

File: app/controllers/list_responses_controller.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
+class ListResponsesController < ApplicationController
+  def feedbacks
+    authorize :list_response, :feedbacks?
```
Comment (toupeira) [inline app/controllers/list_responses_controller.rb]: The second argument to `authorize` is optional here since it already defaults to the action name with `?` appended (same for the other actions)

File: app/controllers/list_responses_controller.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  def hours
+    authorize :list_response, :hours?
+    @hours = Hour.all
+  end
+
+  def mark_feedback_done
+    feedback = Feedback.find(params[:id])
+    authorize :list_response, :mark_feedback_done?
+    feedback.update(marked_done_by: current_user)
+    redirect_to list_responses_feedbacks_path, notice: 'Erfolgreich erledigt markiert'
```
Comment (toupeira) [inline app/controllers/list_responses_controller.rb]: "Erfolgreich *als* erledigt markiert"

Also, do we need a redirect and/or flash notice in the other `mark_*` actions?

File: app/models/assignment.rb [hunk 4]
```diff
@@ -65,9 +65,17 @@ def delete_reminder
   end
 
   def to_label
-    label = I18n.t('activerecord.models.assignment')
-    label += " - #{client.contact.full_name}" if client.contact.present?
-    label += " - #{period_start && I18n.l(period_start)}"
+    label = label_parts[0]
+    label += " - #{label_parts[1]}" if label_parts[1].present?
+    label += " - #{label_parts[2]}" if label_parts[2].present?
     label
```
Comment (toupeira) [inline app/models/assignment.rb]: `label_parts.compact.join(' - ')` maybe?

File: app/models/group_assignment.rb [hunk 5]
```diff
@@ -38,11 +38,19 @@ def save_group_assignment_logs
   end
 
   def to_label
-    label = "Gruppenangebot - #{group_offer.title}"
-    label += " - #{group_offer.department.contact.last_name}" if group_offer.department.present?
+    label = "#{label_parts[0]} - #{label_parts[1]}"
+    label += " - #{label_parts[2]}" if label_parts[2].present?
```
Comment (toupeira) [inline app/models/group_assignment.rb]: `label_parts.compact.join(' - ')`

File: app/views/application/_modal_truncate.html.slim [hunk 6]
```diff
@@ -0,0 +1,10 @@
+#truncate-modal.modal.fade aria-labelledby="truncate-modal-title" role="dialog" tabindex="-1"
+  .modal-dialog.modal-sm role="document"
+    .modal-content
+      .modal-header.modal-header-sm
+        button.close aria-label="Close" data-dismiss="modal" type="button"
```
Comment (toupeira) [inline app/views/application/_modal_truncate.html.slim]: "Schliessen"

File: config/initializers/mime_types.rb [hunk 7]
```diff
@@ -2,3 +2,4 @@
 
 # Add new mime types for use in respond_to blocks:
 # Mime::Type.register "text/richtext", :rtf
+Mime::Type.register 'application/pdf', :pdf
```
Comment (toupeira) [inline config/initializers/mime_types.rb]: Looks like this already exists:

```
[1] pry » Mime[:pdf]
=> #<Mime::Type:0x000056141c78e020 @synonyms=[], @symbol=:pdf, @string="application/pdf", @hash=-1686108550264282935>
```

File: test/policies/list_response_policy_test.rb [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+  end
+
+  def test_create
+  end
+
+  def test_update
+  end
+
+  def test_destroy
+  end
```
Comment (toupeira) [inline test/policies/list_response_policy_test.rb]: Add tests or remove file

File: app/models/assignment.rb (hunk 4 above)
Comment (casaper) [reply app/models/assignment.rb]: 🤦‍♂️

File: config/initializers/mime_types.rb (hunk 7 above)
Comment (casaper) [reply config/initializers/mime_types.rb]: I did this after I discovered some strange errors in the console running rails, when I was actually trying to resolve the unbelievably slow sass processing. Googling the error message has resolved it being related to this somehow. And since I've put it there I've never seen it again. I presume it wasn't always set before?

File: test/policies/list_response_policy_test.rb (hunk 8 above)
Comment (casaper) [reply test/policies/list_response_policy_test.rb]: actually this is all WIP, so that may explain some of the unfinished stuff.
But still highly appreciate your review 👍 👍

File: config/initializers/mime_types.rb (hunk 7 above)
Comment (casaper) [reply config/initializers/mime_types.rb]: The SASS pain was eased by replacing sass_rails with sassc_rails. Had I only known that earlier, it'd saved loads of 👋 💥 ☠️ 💣 . Changing stuff in SASS could literally use 40 seconds or more to recompile, while with sassc the difference to having made changes is hardly noticable.

File: app/views/application/_modal_truncate.html.slim (hunk 6 above)
Comment (casaper) [reply app/views/application/_modal_truncate.html.slim]: Makes sense.

File: app/assets/javascripts/volunteer_form.es6 [hunk 9]
```diff
@@ -1,28 +1,23 @@
-$(function() {
-  $( document ).on('turbolinks:render, turbolinks:load', function() {
-    show_rejection();
-    show_experience();
```
Comment (chrysalag) [inline app/assets/javascripts/volunteer_form.es6]: Why is this removed?

File: app/views/application/_navigation.html.slim [hunk 10]
```diff
@@ -11,31 +11,32 @@ nav.navbar.navbar-top.hidden-print
...
-            li = link_to 'Erinnerungs-Emails', reminder_mailings_path
         ul.nav.navbar-nav.navbar-right
           li.dropdown#menu
             a.dropdown-toggle aria-expanded='false' aria-haspopup='true' data-toggle='dropdown' href='#' role='button'
-              span= current_user.navigation_name
-              span= t("role.#{current_user.role}")
               span.caret
+              span.user-account
+                span= t("role.#{current_user.role}")
+                span= current_user.email
```
Comment (chrysalag) [inline app/views/application/_navigation.html.slim]: span= current_user.full_name (requirement from client)

File: app/views/list_responses/feedbacks.html.slim [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+          span= "#{(Time.zone.today - record.created_at.to_date).to_i} Tage"
+          - if record.reviewer.present?
+            span
+              = "Angeschaut von: "
+              = link_to record.reviewer.email, record.reviewer
+        td
+          - if record.reviewer.present?
+            = link_to 'Anzeigen', polymorphic_path([record.volunteer, record.feedbackable_link_target, record]), class: 'index-action-link'
+          - else
+            = link_to 'Angeschaut', polymorphic_path([record.volunteer, record.feedbackable_link_target, record], action: :mark_as_done, q: search_parameters), { method: :put, class: 'btn btn-success btn-xs m-t-10' }
```
Comment (chrysalag) [inline app/views/list_responses/feedbacks.html.slim]: `btn-success` should be replaced with `btn-default`. They don't want colors in buttons except for the blue Journal buttons. (Changes regarding all other buttons are already merged.)

File: app/views/list_responses/hours.html.slim [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+          span #{(Time.zone.today - record.created_at.to_date).to_i} Tage
+          - if record.reviewer.present?
+            span
+              = "Angeschaut von: "
+              = link_to record.reviewer.email, record.reviewer
+        td
+          - if record.reviewer.present?
+            = link_to 'Anzeigen', polymorphic_path([record.volunteer, record.hourable, record]), class: 'index-action-link'
+          - else
+            = link_to 'Angeschaut', polymorphic_path([record.volunteer, record.hourable, record], action: :mark_as_done, q: search_parameters), { method: :put, class: 'btn btn-success btn-xs m-t-10' }
```
Comment (chrysalag) [inline app/views/list_responses/hours.html.slim]: Same here.

File: app/views/list_responses/trial_feedbacks.html.slim [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+          span= "#{(Time.zone.today - record.created_at.to_date).to_i} Tage"
+          - if record.reviewer.present?
+            span
+              = "Angeschaut von: "
+              = link_to record.reviewer.email, record.reviewer
+        td
+          - if record.reviewer.present?
+            = link_to 'Anzeigen', polymorphic_path([record.volunteer, record.trial_feedbackable_link_target, record]), class: 'index-action-link'
+          - else
+            = link_to 'Angeschaut', polymorphic_path([record.volunteer, record.trial_feedbackable_link_target, record], action: :mark_as_done, q: search_parameters), { method: :put, class: 'btn btn-success btn-xs m-t-10' }
```
Comment (chrysalag) [inline app/views/list_responses/trial_feedbacks.html.slim]: Same here.

File: app/views/volunteers/_form.html.slim [hunk 14]
```diff
@@ -64,8 +64,8 @@ fieldset
   .col-xs-12
     = f.input :motivation
   .col-xs-12
-    = f.input :experience
-  .col-xs-12
+    = checkbox_toggle_collapse(f, :experience, '#experience-desc')
```
Comment (chrysalag) [inline app/views/volunteers/_form.html.slim]: Nice change, not part of the story though.

Comment (casaper) [reply app/views/volunteers/_form.html.slim]: I know, but as the test for this was failing I had to fix it. And I'd rather fix it right instead of keeping it as wrecked. So if we consider fixing tests as part of the story, it is... 😉

File: app/models/concerns/feedback_trial_feedback_common.rb [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+        all
+      end
+    }
+
+    scope :author_isnt_volunteer, (-> { joins(:volunteer).where('author_id != volunteers.user_id') })
+    scope :author_is_volunteer, (-> { joins(:volunteer).where('author_id = volunteers.user_id') })
+
+    scope :since_last_submitted, lambda { |submitted_at|
+      if submitted_at
+        author_volunteer.submitted_before(submitted_before)
```
Comment (chrysalag) [inline app/models/concerns/feedback_trial_feedback_common.rb]: `author_volunteer.submitted_before(submitted_at)` ?

File: app/models/volunteer.rb [hunk 16]
```diff
@@ -6,11 +6,9 @@ class Volunteer < ApplicationRecord
   include ImportRelation
   include FullBankDetails
 
-  before_validation :handle_external
+  before_validation :handle_user_with_external_change, if: :external_changed?
+  # before_validation :handle_external
```
Comment (chrysalag) [inline app/models/volunteer.rb]: comment that needs to be removed

File: app/views/list_responses/trial_feedbacks.html.slim [hunk 17]
```diff
@@ -0,0 +1,57 @@
+= render 'reminder_mailings/section_navigation'
+
+h1 Eingereichte Probezeit Feedbacks
```
Comment (chrysalag) [inline app/views/list_responses/trial_feedbacks.html.slim]: Then the need_review is no longer needed (along with policy, action, routes etc)

File: app/views/list_responses/feedbacks.html.slim [hunk 18]
```diff
@@ -0,0 +1,66 @@
+= render 'reminder_mailings/section_navigation'
+
+h1 Eingereichte Feedbacks
+
+nav.navbar.section-navigation.hidden-print
+  ul.list-inline
+    li= button_link 'Filter aufheben', url_for(q: { s: ['updated_at asc'] }), dimension: :sm
```
Comment (chrysalag) [inline app/views/list_responses/feedbacks.html.slim]: AFAIK, they only requested to have lists of the unreviewed feedbacks, unreviewed trial feedbacks and unpaid hours.

File: app/views/list_responses/feedbacks.html.slim [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+          a href=volunteer_path(record.volunteer)
+            span #{record.volunteer.contact.last_name},
+            span= record.volunteer.contact.first_name
+        td.index-action-cell
+          a href="#{url_for(record.feedbackable)}" title="#{record.feedbackable.to_label}"
+            - record.feedbackable.label_parts.reject(&:blank?).each do |label_part|
+              span= label_part.truncate(30)
+        td
+          - if record.author.volunteer?
+            = link_to record.volunteer.freiwillig_gendered, record.volunteer
```
Comment (chrysalag) [inline app/views/list_responses/feedbacks.html.slim]: I believe it would be preferable to display the author name.

File: app/views/list_responses/feedbacks.html.slim [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+          - else
+            = link_to record.author.profile.contact.natural_name, record.author
+        td= record.goals
+        = td_truncate_content_modal(record.achievements, 'Erfolge', shorten_size: 60)
```
Comment (chrysalag) [inline app/views/list_responses/feedbacks.html.slim]: They asked for the display of the whole text in journals. If we are to restrict the size of the texts shown, we could at least have an increased limit.

File: app/views/list_responses/feedbacks.html.slim [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+        = td_truncate_content_modal(record.future, 'Zukunft', shorten_size: 60)
+        = td_truncate_content_modal(record.comments, t_attr(:comments, Feedback), shorten_size: 60)
+        td= t(record.conversation)
+        td.index-action-cell
+          span= "#{(Time.zone.today - record.created_at.to_date).to_i} Tage"
```
Comment (chrysalag) [inline app/views/list_responses/feedbacks.html.slim]: We have no feedback that they prefer to see how many days old are the feedbacks.

Comment (casaper) [reply app/views/list_responses/feedbacks.html.slim]: They requested to see a list of feedbacks, oldest first. In order to process the oldest first to have the volunteers not to be waiting. So as It wasn't even specified to display the date at all, I don't see how you can conclude that displaying the age in days instead of the date would be any wronger than the other!?!?

File: app/views/reminder_mailings/_section_navigation.html.slim [hunk 22]
```diff
@@ -1,5 +1,9 @@
 nav.navbar.section-navigation.hidden-print
   ul.section-nav-ul.list-unstyled
-    li= button_link 'Erinnerungs-Email Übersicht', reminder_mailings_path, dimension: :sm
-    li= button_link 'Probezeit Erinnerung erstellen', new_probation_period_reminder_mailings_path, dimension: :sm
-    li= button_link 'Halbjahres Erinnerung erstellen', new_half_year_reminder_mailings_path, dimension: :sm
+    li= section_nav_button('feedbacks', 'Feedback Eingang', list_responses_feedbacks_path(q: default_list_response_query))
```
[the remark this answers is not in the document]
Comment (casaper) [reply app/views/reminder_mailings/_section_navigation.html.slim]: Because it is about processing incomming feedback, just like Posteingang in an Email client for example.

[the remark this answers is not in the document]
Comment (casaper) [reply app/views/reminder_mailings/_section_navigation.html.slim]: Übersicht would be more of feedback in general.

File: app/views/list_responses/trial_feedbacks.html.slim [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+      th= sort_link @q, :author_profile_contact_full_name, 'Autor/in'
+      th Text
+      th= sort_link @q, :created_at, 'Alter'
+      th.limit-width
+  tbody
+    - @trial_feedbacks.each do |record|
+      tr*{ data: { feedback_id: record.id } }
+        td.index-action-cell
+          a href=volunteer_path(record.volunteer)
+            span #{record.volunteer.contact.last_name},
```
[the remark this answers is not in the document]
Comment (casaper) [reply app/views/list_responses/trial_feedbacks.html.slim]: it is full_name, but in a way that makes the column not that wide.

File: app/views/list_responses/hours.html.slim [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+  tbody
+    - @hours.each do |record|
+      tr*{ data: { hour_id: record.id } }
+        td= link_to record.volunteer.full_name, volunteer_path(record.volunteer)
+        td
+          a href="#{url_for(record.hourable)}" title="#{record.hourable.to_label}"
+            = record.hourable.to_label
+        td.index-action-cell
+          span
+            - total_hours = (60.0 / record.minutes * 0.1) + record.hours
```
[the remark this answers is not in the document]
Comment (casaper) [reply app/views/list_responses/hours.html.slim]: no, as 1 hour is 60 minutes that math would be fundamentally wrong.

File: app/models/concerns/feedback_trial_feedback_common.rb [hunk 25]
```diff
@@ -0,0 +1,29 @@
...
+  included do
+    include ReviewsCommon
+
+    scope :author_volunteer, lambda { |toggle|
+      return all unless toggle
+      if toggle[:author_volunteer] == 'true'
+        author_is_volunteer
+      elsif toggle[:author_volunteer] == 'false'
+        author_isnt_volunteer
+      else
```
[the remark this answers is not in the document]
Comment (casaper) [reply app/models/concerns/feedback_trial_feedback_common.rb]: what if its value is present, but neither `false` or `true`. In that case it could possibly lead to the red thing....

File: app/views/list_responses/hours.html.slim (hunk 24 above)
[the remark this answers is not in the document]
Comment (casaper) [reply app/views/list_responses/hours.html.slim]: well, I guess both of our maths is wrong. I'll fix it ...
